### ISSUE TITLE
Equidistant nodes test fail from time to time

### ISSUE THREAD

**Issue by pancetta** [Parallel-in-Time/pySDC#147]
For some numbers, the tests using the new node generator leads to errors when testing the equidistant distribution. Looks like it happens only for some random numbers and is thus hard to reproduce. One instance of this can be seen here: <https://github.com/Parallel-in-Time/pySDC/runs/7713874352?check_suite_focus=true#step:5:339>

@tlunet: If you have an idea, go ahead and fix. If not, then I'll fix the random seed and ignore it.

**Comment by tlunet** [comment]
Hum ... seems to be the spline interpolation that is failing, rather than the collocation test of pySDC. I don't know much about spline interpolation, but from the error message it seems that the right bound interval xe (=array([0.81757221])) is not larger than the last grid point x[m-1], with x = array([0.47713716, 0.81757221]). From the look of it, seems to be some small error due to machine precision, that make xe < x[m-1] = x[-1].

After looking a bit, I think this comes probably from the re-scaling done in Collocation._getNodes, after computation of the normalized nodes between [-1, 1] :

https://github.com/Parallel-in-Time/pySDC/blob/b05bff9192658f1ce4784b4a44d74aea7f2e2138/pySDC/core/Collocation.py#L172-L178

For some configurations, with RADAU-RIGHT nodes, it seems that the epsilon due to machine precision produces a last point for the nodes that is not equal to self.tright, and some times bigger, which trigger the error in the spline interpolation.

If I'm not wrong with this interpretation, then I think the solution would be to add some security modification at the end of Collocation._getNodes, when the node distribution uses RADAU-RIGHT or LOBATTO :

```python
if self.right_is_node:
    nodes[-1] = self.tright
```

Something equivalent could be done with self.tleft when node distribution uses RADAU-LEFT or LOBATTO :

```python
if self.left_is_node:
    nodes[0] = self.tleft
```

Also, I don't think this should matter since Python automatically cast integer into double precision when doing mixed-type arithmetic operations, but maybe modifying the scaling to :

```python
a = self.tleft
b = self.tright
nodes += 1.0
nodes /= 2.0
nodes *= b - a
nodes += a
```

could have an impact too (but it would be weird ...).

So let me know what you think of this @pancetta, and if this looks good to you, I can make a quick PR for this (even if a small commit in the master branch would be enough for this :smirk: )

**Comment by tlunet** [comment]
PS : just saw that the test fails for python 3.10, and not for the previous versions. So maybe the modification for the re-scaling is important ... not sure about the changes brought by python 3.10 for mixed-type floating point operations ...

**Comment by pancetta** [comment]
I agree, this is a random thing and not a version thing. I added a random seed [here](https://github.com/Parallel-in-Time/pySDC/commit/90053a09ef57111a683f237414b1966b4ee9581a) and we should be fine. Thanks for the explanation!

**Comment by tlunet** [comment]
Yeah, I saw that commit for v5 (are you 42 years old ? ... just wondering where this 42 comes from :smile:). 

Personally, I would suggest to do the safety and the rescaling modifications : it doesn't cost much, and in that case it's better to be safe than sorry. Then I guess the fixed random seed could be removed, and future experience will confirm (or not) if this was the right choice :sweat_smile:

**Comment by pancetta** [comment]
Sounds good to me. Want to do a PR?

**Comment by tlunet** [comment]
Only for this ? It would be more simple to do a commit on the v5 branch. I can maybe do it myself if I have writing right on this branch (this would be the case with gitlab, not so sure how it works in github)

**Comment by pancetta** [comment]
v5 will be merged to the master at some point. Besides on PR from @ikrom96git, v5 is up-to-date with master anyway. I'll clean up a few things in v5 and then merge back to master, make a new release of pySDC and then the game restarts.

**Comment by tlunet** [comment]
so, commit on v5 or PR for master, that is still my question (between the two, my heart cannot choose)

**Comment by pancetta** [comment]
It does not matter at the end, that's what I wanted to say above. You could also do a PR to v5. If you don't I'll just do a commit to v5 on the way.

**Comment by tlunet** [comment]
Ok, then I'll do a PR on v5 soon

**Comment by pancetta** [comment]
Nice, thanks!

**Comment by pancetta** [comment]
Closed by #151, thanks @tlunet!
